File: 35-search-insert-position/35-search-insert-position.java
class Solution {
    public int searchInsert(int[] nums, int target) {
        // One pass binary search
        int high = nums.length - 1;
        int low = 0;
        int middle = 0;
        
        while (low <= high){
            middle = (low + (high - low) / 2);
            if (target < nums[middle]){
                high = middle - 1;
            }
            else if (target > nums[middle]){
                low = middle + 1;
            }
            else if (target == nums[middle]){
                return middle;
            }
        }
        return low;
    }
}

File: 35-search-insert-position/35-search-insert-position.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        while (left <= right):
            pivot = (left + right) // 2
            if (target == nums[pivot]):
                return pivot
            if (target < nums[pivot]):
                right = pivot - 1
            if (target > nums[pivot]):
                left =  pivot + 1
        return left